feat: (Platform) introduce new Object List Item Component

sKudum opened the pull request:
#### Please provide a link to the associated issue.

closes #3484 
closes #3415 
closes #3290 
closes #2357 
closes #2358 

#### Please provide a brief summary of this pull request.

This PR consists of.
1). Object list item.
 
-   Standard structure.
-    Row select-multi
-    Navigation
-    Row selection(Single) with Navigation.

2).Action List Item
   - +View more Key board support. on Enter and space.

3). Display List Item
    - + With partial Navigation feature.

4).List.
   - Key up and down arrow proper flow fix.

#### Please check whether the PR fulfills the following requirements

- [x] the commit message follows the guideline:
https://github.com/SAP/fundamental-ngx/blob/master/CONTRIBUTING.md
- [x] tests for the changes that have been done
- [x ] all items on the PR Review Checklist are addressed :
https://github.com/SAP/fundamental-ngx/wiki/PR-Review-Checklist

Documentation checklist:
- [x] Documentation Examples
- [x] Stackblitz works for all examples

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-action-list-item/platform-action-list-item-examples/platform-action-list-item-border-less-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+        { 'title': 'Action 4' }];
+
+    @HostListener('click', ['$event'])
+    _deleteRow(event: any): void {
```
plz add type not use any

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-action-list-item/platform-action-list-item-examples/platform-action-list-item-border-less-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+    }
+
+    @HostListener('keyup', ['$event'])
+    _onKeydown(event: any): void {
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-action-list-item/platform-action-list-item-examples/platform-action-list-item-border-less-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+
+    @HostListener('keyup', ['$event'])
+    _onKeydown(event: any): void {
+        if (event && (event.key === 'Enter' || event.key === 'Space')) {
```
plz use here KeyUtil with @angular/cdk/keycodes

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-action-list-item/platform-action-list-item-examples/platform-action-list-item-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+    templateUrl: './platform-action-list-item-example.component.html'
+})
+export class PlatformActionListItemExampleComponent {
+    items: any[] = [
```
the same comments for whole file

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-display-list-item/platform-display-list-item-examples/platform-display-list-item-border-less-example.component.ts:
```diff
@@ -0,0 +1,14 @@
+    templateUrl: './platform-borderless-display-list-item-example.component.html'
+})
+export class PlatformDisplayListItemBorderLessExampleComponent {
+    items: any[] = [
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-display-list-item/platform-display-list-item-examples/platform-display-list-item-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+    templateUrl: './platform-display-list-item-with-navigation-example.component.html'
+})
+export class PlatformDisplayListItemWithNavigationExampleComponent {
+    items: any[] = [
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-object-list-item/platform-object-list-item-examples/platform-object-list-item-border-less-example.component.ts:
```diff
@@ -0,0 +1,93 @@
+    templateUrl: './platform-borderless-object-list-item-example.component.html'
+})
+export class PlatformObjectListItemBorderLessExampleComponent {
+    items: any[] = [
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-object-list-item/platform-object-list-item-examples/platform-object-list-item-with-row-navigation-example.component.ts:
```diff
@@ -0,0 +1,145 @@
+})
+export class PlatformObjectListItemWithRowNavigationExampleComponent {
+    _dataSource = new ListDataSource<Product>(new ListDataProvider());
+    _selectedItems: any[] = [];
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-object-list-item/platform-object-list-item-examples/platform-object-list-item-with-row-selection-and-navigation-example.component.ts:
```diff
@@ -0,0 +1,137 @@
+})
+export class PlatformObjectListItemWithRowSelectionAndNavigationExampleComponent {
+    _dataSource = new ListDataSource<Product>(new ListDataProvider());
+    _selectedItems: any[] = [];
```
the same

Betrozov commented on apps/docs/src/app/platform/component-docs/platform-object-list-item/platform-object-list-item-examples/platform-object-list-item-with-row-selection-example.component.ts:
```diff
@@ -0,0 +1,135 @@
+import { ListDataSource, DataProvider } from '@fundamental-ngx/platform';
+
+
+const LIST_ELEMENTS: Product[] = [
```
please move to it to the own file and import it everywhere

sKudum replied:
Each example has different values with different scenarios.

Betrozov commented on libs/platform/src/lib/components/list/action-list-item/action-list-item.component.ts:
```diff
@@ -0,0 +1,46 @@
+    /** @hidden */
+    /**on keydown append active styles on actionable item */
+    _onKeyDown(event: any): void {
+        if (event.keyCode === ENTER || event.keyCode === SPACE) {
```
plz use here KeyUtil with @angular/cdk/keycodes

Betrozov commented on libs/platform/src/lib/components/list/action-list-item/action-list-item.component.ts:
```diff
@@ -0,0 +1,46 @@
+    /** @hidden */
+    /**on keyup remove active styles from actionable item*/
+    _onKeyUp(event: any): void {
+        if (event.keyCode === ENTER || event.keyCode === SPACE) {
```
the same

Betrozov commented on libs/platform/src/lib/components/list/action-list-item/action-list-item.component.ts:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    /** @hidden */
+    constructor(_changeDetectorRef: ChangeDetectorRef, public itemEl: ElementRef,
```
pl remove public
public by default

Betrozov commented on libs/platform/src/lib/components/list/base-list-item.ts:
```diff
@@ -172,6 +227,21 @@ export class BaseListItem extends BaseComponent implements OnInit, AfterViewChec
     /**Getter and setter the list of items */
     @Input()
     set item(item: any) {
```
plz add type

fkolar replied:
Dont use `item:any` I assume you already know the data type you are working

Betrozov commented on libs/platform/src/lib/components/list/display-list-item/display-list-item.component.ts:
```diff
@@ -0,0 +1,25 @@
+})
+export class DisplayListItemComponent extends BaseListItem {
+    /** @hidden */
+    constructor(_changeDetectorRef: ChangeDetectorRef, public itemEl: ElementRef,
```
the same

Betrozov commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -359,20 +417,40 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
      * set values when passed via datasource
      */
     ngAfterContentInit(): void {
+        setTimeout(() => {
```
why do you need here setTimeout?
if it really needs plz use rxjs timer

Betrozov commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -359,20 +417,40 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
+            }
+        })
         this._itemsSubscription = this.ListItems.changes.subscribe((items) => {
+            setTimeout(() => {
```
the same

Betrozov commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -584,14 +679,66 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
+        // handles mutli select on row level without checkbox
+        if (this.selectionMode === 'multi' && this.selectRow) {
+            this.ListItems.forEach((item) => {
+                if (item._selected) {
```
```suggestion
                if (item._selected) {
                    this._selectionModel.select(item);
                } else {
                    this._selectionModel.deselect(item);
                }
                 this.stateChanges.next(item);
```

Betrozov commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -418,6 +496,14 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
+    /** @hidden */
+    /**load more on enter or space press */
+    _loadOnkeyPress(event: KeyboardEvent): void {
+        if (event.keyCode === ENTER || event.keyCode === SPACE) {
```
plz use KeyUtil

Betrozov commented on libs/platform/src/lib/components/list/object-list-item/object-list-item.component.ts:
```diff
@@ -0,0 +1,40 @@
+    isCompact = this._contentDensity === 'compact';
+
+    /** @hidden */
+    constructor(_changeDetectorRef: ChangeDetectorRef, public itemEl: ElementRef,
```
the same

Betrozov commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -359,20 +417,40 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
      */
     ngAfterContentInit(): void {
+        setTimeout(() => {
+            if (this.ListItems.length !== 0) {
```
plz change `ListItems ` to `listItems `

KevinOkamoto commented on libs/platform/src/lib/components/list/action-list-item/action-list-item.component.ts:
```diff
@@ -0,0 +1,46 @@
+
+    @Input()
+    title: string;
+
```
Does this component emit an event on button click? If it doesn't we should add that feature.

KevinOkamoto reviewed (changes requested):
See comments.

sKudum commented on libs/platform/src/lib/components/list/base-list-item.ts:
```diff
@@ -172,6 +227,21 @@ export class BaseListItem extends BaseComponent implements OnInit, AfterViewChec
     /**Getter and setter the list of items */
     @Input()
     set item(item: any) {
```
as per my understanding, A 'set' accessor cannot have a return type annotation, it gives error even tho we add it.

sKudum commented on libs/platform/src/lib/components/list/object-list-item/object-list-item-row.component.ts:
```diff
@@ -0,0 +1,38 @@
+            <fdp-object-attribute class="fd-object-list__object-attribute"
+             [islink]="islink" [label]="attributeLabel"
+              [linkText]="linkedText">
+              <ng-content select="#attribute"></ng-content>
```
it is for setting special charters such as :  between text and linked text of object attribute.

KevinOkamoto replied:
Is there a reason for using an `id` (i.e. `#attribute`) as the selector for `ng-content`? Normally HTML `id` should be unique values on a web page, and since this is in a list item it's appearance may show up multiple times.

It would be better to use a class selector (`.attribute`), or an attribute selector, e.g. `[slot='attribute']`.

sKudum commented on libs/platform/src/lib/components/list/base-list-item.ts:
```diff
@@ -39,6 +43,58 @@ interface SecondaryActionItem {
+    /** object number amount decimal limitation */
+    @Input()
+    decimal: number;
+
+    /** object number status */
+    @Input()
+    numberStatus: StatusType;
+
+    /** holds avatar label */
+    @Input()
+    label: string;
+
+    /** holds avatar icon as image */
+    @Input()
+    glyph: string;
+
+    /** Avatar to be circle or not */
+    @Input()
+    circle: boolean;
+
+    /** Is avatar has placeholder */
+    @Input()
+    placeholder: boolean;
+
+    /** Is avatar has tile */
+    @Input()
+    tile: boolean;
+
+    /** Is avatar has colorAccent */
+    @Input()
+    colorAccent: number;
+
+    /** Avatar image path */
+    @Input()
+    image: string;
+
+    /** Avatar is transparent */
+    @Input()
+    transparent: boolean;
+
```
As discussed, segregation is leading to breaking the imperative approach.If possible will show you an example for it tomorrow in loca.

dimamarksman reviewed (changes requested):
Mostly questions

dimamarksman commented on apps/docs/src/app/platform/component-docs/platform-action-list-item/platform-action-list-item-examples/platform-action-list-item-example.component.ts:
```diff
@@ -0,0 +1,27 @@
+        { 'title': 'Action 3' },
+        { 'title': 'Action 4' }];
+
+    @HostListener('click', ['$event'])
+    _deleteRow(event: any): void {
+        if (event.target.tagName.toLowerCase() === 'button') {
+            alert('invoked :' + event.target.getAttribute('title'));
+        }
+    }
+
+    @HostListener('keyup', ['$event'])
+    _onKeydown(event: any): void {
+        if (event && (event.key === 'Enter' || event.key === 'Space')) {
+            alert('invoked :' + event.target.parentNode.getAttribute('title'));
+        }
+    }
```
Sorry, this example is not about contentDensity.
But anyway do we really need to expose this code to the users in the example?

dimamarksman commented on apps/docs/src/app/platform/component-docs/platform-display-list-item/platform-display-list-item-header/platform-display-list-item-header.component.ts:
```diff
@@ -0,0 +1,7 @@
+import { Component } from '@angular/core';
```
sorry, I meant `apps/docs/src/app/platform/component-docs/platform-display-list-item/platform-display-list-item-header/platform-display-list-item-header.component.scss`

dimamarksman commented on libs/platform/src/lib/components/list/base-list-item.ts:
```diff
@@ -304,11 +384,22 @@ export class BaseListItem extends BaseComponent implements OnInit, AfterViewChec
                 this._onKeyboardClick(event);
             }
+            if (this.anchor) {
+                const path = this.anchor.nativeElement.getAttribute('ng-reflect-router-link');
```
@sKudum you can try this:
```
ViewChild(RouterLink) anchor: RouterLink;
...
this.anchor.onClick(); // Trigger link navigation manually
```

sKudum commented:
Hi @fkolar , @dimamarksman , @dimamarksman , @KevinOkamoto .
Please have a look all your comments are addressed

1). Removed any[] and defined type.
2).event:any is changed to the custom event.
3).settimeout is removed.
4).removed empty files.
5). Removed unused imports.
6).defined custom event for action list item on click.
7).secondary actions as replaced with <ng-content>
8).updated examples alert message display, anchor.click as suggested.

Note: During these changes, I found in older files also I can enhance such as adding a custom event on the item selected, update navigation, will handle these separately, as this PR is targeting for the new components.

dimamarksman commented on libs/platform/src/lib/components/list/object-list-item/object-list-item.component.html:
```diff
@@ -0,0 +1,92 @@
+        class="fd-object-list__item fd-list__item--link" role="listitem"
+        [class.fd-list__item--inactive]="listType === 'inactive'" [attr.ariaLabelledBy]="ariaLabelledBy"
+        [attr.aria-level]="ariaLevel" [attr.aria-posinet]="ariaPosinet">
+        <ng-container *ngTemplateOutlet=" selectionTemplate">
```
please remove empty space

fkolar reviewed (approved):
Looks ok I would probably rename a few @Inputs, but its ok.

JKMarkowski commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -418,6 +492,14 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
+    /** @hidden */
+    /**load more on enter or space press */
+    _loadOnkeyPress(event: KeyboardEvent): void {
+        if (KeyUtil.isKeyCode(event, SPACE) || KeyUtil.isKeyCode(event, SPACE)) {
```
Redundant statement

JKMarkowski commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -459,19 +541,26 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
+     * event:any to avoid code duplication
+     * seprate PR for custom event**/
     @HostListener('click', ['$event'])
     _updateNavigation(event: any): void {
```
Can you change any to MouseEvent?

JKMarkowski commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -576,22 +668,73 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
 
+
+
+    private _handleRowSelect(event): void {
```
Can you add a type there?

JKMarkowski commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -535,8 +624,11 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
      */
     private _handleSingleSelect(event: any): void {
```
Can you change this any to Event type?

JKMarkowski commented on libs/platform/src/lib/components/list/list.component.ts:
```diff
@@ -576,22 +668,73 @@ export class ListComponent extends CollectionBaseInput implements OnInit, AfterV
     /**List item with checkbox styles,check,uncheckupdates
      * event:any to avoid code duplication
      */
     private _handleMultiSelect(event: any): void {
```
Can you change this any to Event type?

JKMarkowski reviewed (approved):
Hello @sKudum I left some minor comments, rest Looks good to me, so I will approve it
